### PR TITLE
Run buildkit integration tests

### PR DESCRIPTION
**Pull request by tiborvass**
Signed-off-by: Sam Whited <sam@samwhited.com>

### CODE REVIEW

File: Jenkinsfile
Comment (tiborvass) [inline Jenkinsfile]: @SamWhited so yeah all you need here is to prepend `DOCKER_BUILDKIT=1`

File: hack/make/test-buildkit
Comment (thaJeztah) [inline hack/make/test-buildkit]: Instead of using `echo` here, you can use this to show the commands that are being run if the `TESTDEBUG` env-var is set;

```bash
[ -n "${TESTDEBUG}" ] && set -x
```

File: hack/make/test-buildkit
Comment (thaJeztah) [inline hack/make/test-buildkit]: Same here;

```suggestion
	tar -xzf "${PWD}"/buildkit.tar.gz
```

File: hack/make/test-buildkit
Comment (thaJeztah) [inline hack/make/test-buildkit]:
```suggestion
	mv buildkit-"${BUILDKIT_COMMIT}"/ "${GOPATH}"/buildkit/
```

The curly brackets are redundant, but it's good to be consistent

(there's some more variables that could use quotes and/or curly brackets below, I won't point them out to reduce the noise a bit)

File: Jenkinsfile
Comment (thaJeztah) [inline Jenkinsfile]: A bit of a "nit", but in hindsight, perhaps I should've named `skip_dco` `dco` for consistency with other stages. In line with that, name this `buildkit`, and enable it by default?

```suggestion
        booleanParam(name: 'buildkit', defaultValue: true, description: 'buildkit integration tests')
```

I'm also thinking of sorting this list alphabetically, so that they match with the position in which they're running in the Jenkins UI

File: Jenkinsfile
Comment (thaJeztah) [inline Jenkinsfile]: typo;

```suggestion
        BUILDKIT_COMMIT      = 'abe19c27e7288c46893587a42199c52db106e704'
```

File: hack/make/test-buildkit
Comment (thaJeztah) [inline hack/make/test-buildkit]: (just wondering) do these have to be exported, or would just putting these before `./hack test` do the trick?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
I know this isn't ready for review yet, but left some suggestions

File: hack/make/test-buildkit
Comment (SamWhited) [reply hack/make/test-buildkit]: Done

File: hack/make/test-buildkit
Comment (SamWhited) [reply hack/make/test-buildkit]: Done

File: Jenkinsfile
Comment (SamWhited) [reply Jenkinsfile]: Done

File: Jenkinsfile
Comment (SamWhited) [reply Jenkinsfile]: oops, wonder when that happened. Fixed, thanks!

File: hack/make/test-buildkit
Comment (SamWhited) [reply hack/make/test-buildkit]: They'll work before `./hack test`, I think I just moved them so the line wasn't as long, but it doesn't really matter and I can use line continuation for that. Will move them back: tightly limiting the scope is more important than minor style tweaks.

Comment (thaJeztah) [reply hack/make/test-buildkit]: Actually; given that it runs in a sub shell, I don't think scope is an issue (but won't hurt either) 🤷‍♂

File: hack/make/test-buildkit
Comment (thaJeztah) [inline hack/make/test-buildkit]: I think the `./hack/test integration` will build, and run an image, correct?

Not top-priority, but to keep things more separate, wondering if we should use a similar approach as for the `docker-py` tests, where we build the Dockerfile directly from their repository; https://github.com/moby/moby/blob/5a703ccb469c1bd2c75d80ed9891c432078db316/hack/make/test-docker-py#L54

And then run them; https://github.com/moby/moby/blob/5a703ccb469c1bd2c75d80ed9891c432078db316/hack/make/test-docker-py#L62


The problem with that test is still that we build (and run) it docker-in-docker; I think we should run it on the host, and;

1. build the buildkit (or docker-py) image
2. start a "docker-dev" container, which runs the daemon
3. start the buildkit (or docker-py) image, and give it access to the test daemon running in the docker-dev container
4. run the tests


Step 1. could be skipped if BuildKit would push an integration-test image to Docker Hub, so that we only need to pull and run it

Comment (SamWhited) [reply hack/make/test-buildkit]: > I think the ./hack/test integration will build, and run an image, correct?

That's right

> Not top-priority, but to keep things more separate, wondering if we should use a similar approach as for the docker-py tests, where we build the Dockerfile directly from their repository; 

I hadn't seen this, but I like it much better, thanks! I'll think about how this can be restructured.

I'd love to get rid of the docker-in-docker stuff too, one of the difficulties I've had tweaking these tests is that my local system always has some slight difference from CI, so if I want to actually see if things are working I more or less have to use CI. Getting rid of some of the complexity would go a long way towards making it easier to test things locally.

Comment (thaJeztah) [reply hack/make/test-buildkit]: > I'd love to get rid of the docker-in-docker stuff too

I don't mind the docker-in-docker for the regular integration tests, but for these, it means we'll be building the same Dockerfile for each run, and no caching whatsoever (because the container in which it's built is removed afterwards)

So, I haven't tried yet to change this for the docker-py test, but it would be something like either a Make target, or in the jenkinsfile, to spin up the two containers; either sharing the `/var/run/docker.sock`, or (perhaps through a compose-file) have them in the same network so that the client can use a tcp connection

Again; just thinking out loud; figuring out a way to keep the separation as much as possible, but still being able to run the tests of "external" project (docker-py, buildkit, perhaps docker-cli) against a locally built daemon

File: Makefile
Comment (tiborvass) [inline Makefile]: binary alone should be enough

File: Makefile
Comment (tiborvass) [inline Makefile]: You are testing the docker daemon that's provided by Jenkins, not the docker daemon in the codebase. If you look at how `test-docker-py` works as an example, that runs inside the container: `$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary test-docker-py` and buildkit should do the same. So you do need a `hack/make/test-buildkit` in moby that will call into buildkit's `/hack/test integration` from within the container.

Comment (SamWhited) [reply Makefile]: The reason this ended up being changed was because the buildkit integration tests also try to launch containers and the docker-in-docker stuff was likely to be a source of problems and confusion, although I don't remember the specific issues I was running into. I'll go reread the conversation with @thaJeztah and play around with reverting it back; thanks!

**Comment by andrewhsu** [comment]
@SamWhited i added b43076e to always generate a bundle tgz for the buildkit step. see https://github.com/moby/moby/pull/40023/commits/b43076e6977cf64dc7465c4ca55325cb2e282870#diff-58231b16fdee45a03a4ee3cf94a9f2c3R113 for how it collects files to put into the tgz.

**Comment by SamWhited** [comment]
:green_heart: tests! Going to work on getting a more recent commit of Docker vendored in buildkit, and then update the buildkit commit used in these tests and re-enable all the other tests now.

**Comment by SamWhited** [comment]
Roadmap:

- Merge https://github.com/moby/buildkit/pull/1245
- Merge https://github.com/moby/buildkit/pull/1164
- Update this PR to use official buildkit repo instead of my fork
- Re-enable all tests
- Merge!

**Comment by SamWhited** [comment]
@tibor can you mark this as ready for review? I don't have permissions. Thanks!

**Comment by thaJeztah** [comment]
moved it out of draft

**Comment by AkihiroSuda** [comment]
What's current status?

**Comment by SamWhited** [comment]
Just needs review and merge, IIRC, all requested changes should have been made or explained.

**Review by AkihiroSuda — APPROVED** [review]
LGTM if green

**Comment by thaJeztah** [comment]
@SamWhited could you rebase the PR to force CI to run again?

File: Dockerfile
Comment (thaJeztah) [inline Dockerfile]: this looks to be v2.7.0-rc.0 https://github.com/docker/distribution/commit/17b3ff188dfde7d5b59a94ab99e6a967b3a59563

Could you use v2.7.1 here (pin by commit: 2461543d988979529609e8cb6fca9ca190dc48da, but add a comment what version it is)

(Also curious; was this needed to make the tests pass?)

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (SamWhited) [reply Dockerfile]: Fixed. I don't remember why this was needed now, sadly.

**Comment by SamWhited** [comment]
As soon as #40814 is merged this can be rebased, pointed at vendor, and then it will also be ready for merge. /cc @tonistiigi

File: hack/make/test-buildkit
Comment (tonistiigi) [inline hack/make/test-buildkit:39]: buildkitd should not be needed

Comment (SamWhited) [reply hack/make/test-buildkit]: Oops. Fixed.

**Comment by SamWhited** [comment]
Tests passing again

**Pull request closed without merging**
